test/test_common/global.cc

```diff
@@ -0,0 +1,70 @@
+#include "test/test_common/global.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Test {
+
+Globals& Globals::instance() {
+  static Globals* h = new Globals;
+  return *h;
+}
+
+std::string Globals::describeActiveSingletonsHelper() {
+  std::string ret;
+  Thread::ReleasableLockGuard map_lock(map_mutex_);
+  for (auto& p : singleton_map_) {
+    std::unique_ptr<Singleton>& singleton = p.second;
+    ASSERT(singleton != nullptr);
+    if (singleton->ptr_ != nullptr) {
+      absl::StrAppend(&ret, "Unexpected active singleton: ", p.first, "\n");
+    }
+  }
+  return ret;
+}
+
+Globals::Singleton& Globals::get(const std::string& type_name, MakeObjectFn make_object) {
+  Thread::ReleasableLockGuard map_lock(map_mutex_);
+  std::unique_ptr<Singleton>& singleton = singleton_map_[type_name];
+
+  if (singleton == nullptr) {
+    // The first time constructing this object, we'll be constructing the
+    // Singleton for the first time, populating it with a fresh instance,
+    // so no need to take the singleton's mutex. But we do need to hold
+    // the map mutex as we are installing the singleton object in the
+    // singleton_map.
+    singleton = std::make_unique<Singleton>(make_object());
+    map_lock.release();
+    return *singleton;
+  }
+
+  // Subsequent accesses will must lock the singleton's mutex to safely populate
+  // .second. However, we can relinquish map_mutex_ as we are no longer
+  // modifying the map; just the .second of the MutexSingleton pointed to by the
+  // map.
```

Comment:
Since this is test-only code, I recommend using a global lock, this will reduce the cognitive burden on code readers/maintainers. Even better, nobody will be tempted to use it in production code! :)

Reply:
Done, good idea.

Reply:
K, should the comment also be updated?